cranelift-codegen/src/simple_preopt.rs

```diff
@@ -536,11 +542,245 @@ fn simplify(pos: &mut FuncCursor, inst: Inst) {
+            arg: cmp_arg,
+            cond: cmp_cond,
+            imm: cmp_imm,
+        } = pos.func.dfg[iconst_inst]
+        {
+            let cmp_imm: i64 = cmp_imm.into();
+            if cmp_imm != 0 {
+                return;
+            }
+
+            // icmp_imm returns non-zero when the comparison is true. So, if
+            // we're branching on zero, we need to invert the condition.
+            let cond = match opcode {
+                Opcode::Brz => cmp_cond.inverse(),
+                Opcode::Brnz => cmp_cond,
+                _ => return,
+            };
+
+            let new_opcode = match cond {
+                IntCC::Equal => Opcode::Brz,
+                IntCC::NotEqual => Opcode::Brnz,
+                _ => return,
+            };
+
+            BranchOptInfo {
+                br_inst: inst,
+                cmp_arg: cmp_arg,
+                args: args.clone(),
+                new_opcode: new_opcode,
+            }
+        } else {
+            return;
+        }
+    } else {
+        return;
+    };
+
+    info.args.as_mut_slice(&mut pos.func.dfg.value_lists)[0] = info.cmp_arg;
+    if let InstructionData::Branch { ref mut opcode, .. } = pos.func.dfg[info.br_inst] {
+        *opcode = info.new_opcode;
+    } else {
+        panic!();
+    }
+}
+
+struct BranchOrderInfo {
+    term_inst: Inst,
+    term_inst_args: ValueList,
+    term_dest: Ebb,
+    cond_inst: Inst,
+    cond_inst_args: ValueList,
+    cond_dest: Ebb,
+    kind: BranchOrderKind,
+}
+
+enum BranchOrderKind {
+    BrzToBrnz(Value),
+    BrnzToBrz(Value),
+    InvertIcmpCond(IntCC, Value, Value),
+}
+
+/// Reorder branches to encourage fallthroughs.
+///
+/// When an ebb ends with a conditional branch followed by an unconditional
+/// branch, this will reorder them if one of them is branching to the next Ebb
+/// layout-wise. The unconditional jump can then become a fallthrough.
+fn branch_order(pos: &mut FuncCursor, cfg: &mut ControlFlowGraph, ebb: Ebb, inst: Inst) {
+    let info = match pos.func.dfg[inst] {
+        InstructionData::Jump {
+            opcode: Opcode::Jump,
+            destination,
+            ref args,
+        } => {
+            let next_ebb = if let Some(next_ebb) = pos.func.layout.next_ebb(ebb) {
+                next_ebb
+            } else {
+                return;
+            };
+
+            if destination == next_ebb {
+                return;
+            }
+
+            let prev_inst = if let Some(prev_inst) = pos.func.layout.prev_inst(inst) {
+                prev_inst
+            } else {
+                return;
+            };
+
+            let prev_inst_data = &pos.func.dfg[prev_inst];
+
+            if let Some(prev_dest) = prev_inst_data.branch_destination() {
+                if prev_dest != next_ebb {
+                    return;
+                }
+            } else {
+                return;
+            }
+
+            match prev_inst_data {
+                InstructionData::Branch {
```

Comment:
Do we want to handle BranchInt (for `brif)` and BranchFloat (for `brff`) here too? (we could do this in a separate PR)

Reply:
I don't *think* those should be necessary here, because they're not supposed to be constructed by frontends and this is a preopt pass. Please correct me if I'm thinking about that incorrectly though.

Reply:
Good point! I asked this since I didn't know these opcodes were private.